### PR TITLE
[C API DOC][DG] C API developer documentation

### PR DESCRIPTION
**Pull request by zhaixuejun1993**
### Details:
 - Add Readme for C API
 - Add sub script about how to create a new application with C API 2.0
 - Add overview about C API (mapping relationship between C++ and C)

### Tickets:
 - 82497

### CODE REVIEW

File: src/bindings/c/docs/OpenVINO_Runtime_C_API_Developer_Guide.md [hunk 1]
```diff
@@ -53,8 +53,26 @@ Supported Python* versions:
...
-[The Mapping between C++ API 2.0 and C API.xlsx](https://github.com/zhaixuejun1993/openvino/files/9792264/The.Mapping.between.C%2B%2B.API.2.0.and.C.API.xlsx)
+| C++ API  | C API  |
+|:---     |:--- |
+| `Core(const std::string& xml_config_file = {})`| `ov_core_create(ov_core_t** core)` or `ov_core_create_with_config(const char* xml_config_file, ov_core_t** core)`|
+|`~Core()`| `ov_core_free(ov_core_t* core)`|
+|`get_versions(const std::string& device_name)`|`ov_core_get_versions_by_device_name(const ov_core_t* core, const char* device_name, ov_core_version_list_t* versions)`|
+|` read_model(const std::string& model_path, const std::string& bin_path = {})`|`ov_core_read_model(const ov_core_t* core, const char* model_path, const char* bin_path, ov_model_t** model)`|
+|`compile_model(const std::shared_ptr<const ov::Model>& model, const std::string& device_name, const AnyMap& properties = {})`|`ov_core_compile_model(const ov_core_t* core, const ov_model_t* model, const char* device_name, const ov_properties_t* property, ov_compiled_model_t** compiled_model)`|
 
+## Roles in development
```
Comment (riverlijunjie) [inline src/bindings/c/docs/OpenVINO_Runtime_C_API_Developer_Guide.md:64]: Rules in development

File: src/bindings/c/docs/OpenVINO_Runtime_C_API_Developer_Guide.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
+# Developer Guide of OpenVINO C API
+
+> **NOTE**: This guide provides a instroduction about OpenVINO C API.
```
Comment (riverlijunjie) [inline src/bindings/c/docs/OpenVINO_Runtime_C_API_Developer_Guide.md:3]:
```suggestion
> **NOTE**: This guide provides instruction about how to develop OpenVINO C API.
```

File: src/bindings/c/docs/OpenVINO_Runtime_C_API_Developer_Guide.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+C API provides simplified interface for OpenVINO functionality that allows to:
+
+- handle the input network models & input/output tensors.
+- load and configure OpenVINO plugins based on device name.
+- perform inference in synchronous and asynchronous modes with arbitrary number of infer requests (the number of infer requests may be limited by target device capabilities).
+
+## Supported OSes
+
+Currently the OpenVINO C API is supported on Ubuntu* 20.04, Microsoft Windows* 10 and CentOS* 7.3 OSes.
```
Comment (riverlijunjie) [inline src/bindings/c/docs/OpenVINO_Runtime_C_API_Developer_Guide.md:12]: MacOS should also be supported.

File: src/bindings/c/docs/OpenVINO_Runtime_C_API_Developer_Guide.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+Supported Python* versions:
```
Comment (riverlijunjie) [inline src/bindings/c/docs/OpenVINO_Runtime_C_API_Developer_Guide.md:13]: C API doesn't depends on python, do we still need list python version information here?

File: src/bindings/c/docs/OpenVINO_Runtime_C_API_Developer_Guide.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+## Roles in development
+* Most of APIs should return "ov_status_e" and the actual output will be put in input/output parameters. 
+* Some specified APIs will return void or boolean per API's funtionality.
+* structure name plus  "_create" suffix, and input/output parameter is "**" format. Such as C++ `ov::Core()` to C `ov_core_create(ov_core_t** core)`.
+* Don't support the vector of hidden structure.
+* As much as possible to apply pass by value  for input parameter of C APIs.
+* Pass by pointer will be used in output parameters
+* Snake_case style, strict unfold of C++ class plus this class method (can add additional action word to make its readable),  the first paramter should be the C structure pointer of the coresponding C++ class. Such as `OPENVINO_C_API(ov_status_e)  <class_name>_<method_name>_<additional_info>(<sturcture_object_pointer>, <input/output parameter>,... );`
+* Don't support class's  template methods.
+* With same name but different parameters, choose the popular one name first, and others will be add edadditional info to distinguish them.
```
Comment (riverlijunjie) [inline src/bindings/c/docs/OpenVINO_Runtime_C_API_Developer_Guide.md:73]:
```suggestion
* With same name but different parameters, choose the popular one name first, and others will be added additional info to distinguish them.
```

File: src/bindings/c/README.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
+# OpenVINO C API
+
+OpenVINO C API is a part of OpenVINO Runtime library.
```
Comment (ilyachur) [inline src/bindings/c/README.md:3]: It is not a part of OpenVINO Runtime library

File: src/bindings/c/README.md [hunk 7]
```diff
@@ -0,0 +1,28 @@
+# OpenVINO C API
+
+OpenVINO C API is a part of OpenVINO Runtime library. 
+This component is responsible for:
+ * Based representations for the objects in OpenVINO - C API provides structures with C code for basical classes inside the OpenVINO, such as "ov_core_t" for "ov::Core". For more information please read [OpenVINO Runtime C API User Guide](./docs/OpenVINO_Runtime_C_API_User_Guide.md).
```
Comment (ilyachur) [inline src/bindings/c/README.md:5]:
```suggestion
 * Based representations for the objects in OpenVINO - C API provides structures with C code for basical classes inside the OpenVINO, such as `ov_core_t` for `ov::Core`. For more information please read [OpenVINO Runtime C API User Guide](./docs/OpenVINO_Runtime_C_API_User_Guide.md).
```

User guides are located on the site. Here we should have only developer documentation

File: src/bindings/c/README.md [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+ * Based operations representation from OpenVINO - C API contains basical interfaces (read model, inference, properity set and so on) with C code. For more information please read [OpenVINO Runtime C API User Guide](./docs/OpenVINO_Runtime_C_API_User_Guide.md).
```
Comment (ilyachur) [inline src/bindings/c/README.md:6]: The second link to `OpenVINO_Runtime_C_API_User_Guide` moreover file has another name

File: src/bindings/c/docs/How_to_Create_C_Sample_with_OpenVINO_C_API.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
+# How to Create C Sample with OpenVINO C API
```
Comment (ilyachur) [inline src/bindings/c/docs/How_to_Create_C_Sample_with_OpenVINO_C_API.md:1]: I don't believe that we need to have this guide. I don't believe that someone will add sample for C API

File: src/bindings/c/docs/How_to_Create_C_Sample_with_OpenVINO_C_API.md [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+## Setting Up the development Environment
+
+To configure the environment for developing with OpenVINO C API, OpenVINO provided script automatically detects the installed C API and configures required environment, run:
+- On Ubuntu: `source <INSTALL_DIR>/setupvars.sh .`
+- On Windows: `setupvars.bat`
+
+For CMakeFile should include C API header file & link library:
+```ruby
```
Comment (ilyachur) [inline src/bindings/c/docs/How_to_Create_C_Sample_with_OpenVINO_C_API.md:10]: Why ruby?

File: src/bindings/c/docs/OpenVINO_Runtime_C_API_Developer_Guide.md (hunk 3 above)
Comment (ilyachur) [reply src/bindings/c/docs/OpenVINO_Runtime_C_API_Developer_Guide.md:12]: I don't think that we need to have this guide. Developers can use different OSs

File: src/bindings/c/docs/OpenVINO_Runtime_C_API_Developer_Guide.md [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+* Function parameter order:  input parameters first and then output parameters.
+* All unchanged input arguments shoud add "const" specificator.
+
+## Tutorials
```
Comment (ilyachur) [inline src/bindings/c/docs/OpenVINO_Runtime_C_API_Developer_Guide.md:77]: Here we have a lot of redundant information please clean up this file. Also I propose to rename it to something like `api_details.md`, `architecture.md` or something else

**Comment by zhaixuejun1993** [comment]
This PR is a draft version. Will continue implement in future.

File: src/bindings/c/README.md [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+ * [include](./include) contains all provided C API headers. Details information about provided API can be found [here](https://docs.openvino.ai/).
+ * [src](./src) contains the implementations of all C APIs.
+ * [tests](./tests) contains all tests for OpenVINO C APIs. More information about OpenVINO C API tests can be found [here](./docs/How_to_Write_Unite_Test.md).
+
+> **NOTE**: Here also includes C API (version 1.0) [header file](./include/c_api/ie_c_api.h), [source file](./src/ie_c_api.cpp), [unite test](./tests/ie_c_api_test.cpp), but version 2.0 is more recommended to users. 
+
+## Tutorials
+
+* [How to integrate OpenVINO C API with Your Application](https://docs.openvino.ai/)
+* [How to wrap OpenVINO objects with C](./docs/How_to_Wrap_OpenVINO_Objects_with_C.md)
```
Comment (ilyachur) [inline src/bindings/c/README.md:82]: nitpick: in all other documents we don't use capital later in the names. I propose to align.

File: src/bindings/c/README.md [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+## How to contribute to the OpenVINO repository
+
+See [CONTRIBUTING](../../../CONTRIBUTING.md) for details. Thank you!
+
+## See also
+
+ * [OpenVINO™ README](../../../README.md)
+ * [Developer documentation](../../../docs/dev/index.md)
+ * [OpenVINO Runtime C API User Guide](https://docs.openvino.ai/)
+ * [Migration of OpenVINO C API](https://docs.openvino.ai/)
```
Comment (ilyachur) [inline src/bindings/c/README.md:97]: Please fix links

File: src/bindings/c/docs/How_to_Debug_C_API_Issues.md [hunk 14]
```diff
@@ -0,0 +1,42 @@
+# How to Debug C API Issues
+
+C API provides exception handling, here is all possible return values of the functions:
+```ruby
```
Comment (ilyachur) [inline src/bindings/c/docs/How_to_Debug_C_API_Issues.md:4]: ruby?

File: src/bindings/c/docs/How_to_Debug_C_API_Issues.md [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+typedef enum {
```
Comment (ilyachur) [inline src/bindings/c/docs/How_to_Debug_C_API_Issues.md:5]: markdown allows to put a reference to real code ([example](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/ir/docs/support_new_opset.md)).
Please reuse this approach

File: src/bindings/c/docs/How_to_Wrap_OpenVINO_Interfaces_with_C.md [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+- Wrap C++ interface to operate object
+- C implement interfaces 
+
+## Wrap C++ interface to create object
+Like the C++ programming create `class` instance, C also need to create related object such as creating `ov::Core`, `ov::Model`, `ov::InferRequest` and so on. C wrap this operation directly and save a shared pointer to the object(C++) back to C `struct` object. Based on the above wrapping method, example about creating `ov::Core` instance will be introduction for more details to illustrate how to wrap C++ interfaces.
+
+Thinking about the C++ interface `Core(const std::string& xml_config_file = {})`, C wrap as following:
+```ruby
+// 1) define C interface name and input parameters
+ov_status_e ov_core_create_with_config(const char* xml_config_file, ov_core_t** core) {
```
Comment (ilyachur) [inline src/bindings/c/docs/How_to_Wrap_OpenVINO_Interfaces_with_C.md:50]: Please use references to real code.

File: src/bindings/c/docs/How_to_Create_C_Sample_with_OpenVINO_C_API.md (hunk 9 above)
Comment (zhaixuejun1993) [reply src/bindings/c/docs/How_to_Create_C_Sample_with_OpenVINO_C_API.md]: Get it. Will remove

File: src/bindings/c/README.md [hunk 17]
```diff
@@ -0,0 +1,97 @@
+# OpenVINO C API
+
+OpenVINO C API is key part to extend OpenVINO for C API users. This component is responsible for:
+ * Representations for the objects in OpenVINO - this part using structure with C code to represent the classes & structures in OpenVINO, such as `ov_core_t` for `ov::Core`. For more details please read [Mapping Relationship of Objects](./docs/Mapping_Relationship_of_Objects.md).
+ * Representations for the API in OpenVINO - this part implemented the most APIs provided by OpenVINO (read model, inference, property set and so on) with C code. For more information please read [Mapping Relationship of Interfaces](./docs/Mapping_Relationship_of_Interfaces.md).
```
[the remark this answers is not in the document]
Comment (zhaixuejun1993) [reply src/bindings/c/README.md]: done

File: src/bindings/c/README.md [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+    paddle--->openvino_c
+    openvino_c--Read paddle---paddle_fe
+
+    ir_fe--->openvino
+    onnx_fe--->openvino
+    paddle_fe--->openvino
+    
+    openvino--infer--->cpu
+    openvino--infer--->gpu
+    openvino--infer--->auto
```
[the remark this answers is not in the document]
Comment (zhaixuejun1993) [reply src/bindings/c/README.md]: done

File: src/bindings/c/README.md [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+
+## Key person
+
+People from the [openvino-c-api-maintainers](https://github.com/orgs/openvinotoolkit/teams/openvino-c-api-maintainers) allows to approve and merge PRs to the C API component. These guys can help in case of any questions about C API component.
+
+## Components
+
+OpenVINO C API has the next structure:
+ * [docs](./docs) contains developer documentations for OpenVINO C APIs.
+ * [include](./include) contains all provided C API headers. Details information about provided API can be found [here](https://docs.openvino.ai/).
```
[the remark this answers is not in the document]
Comment (zhaixuejun1993) [reply src/bindings/c/README.md]: https://docs.openvino.ai/latest/api/api_reference.html will be used.
But I think the C API reference also need to be add here, @ilyachur need some help to trigger doxygen for C

File: src/bindings/c/docs/How_to_Debug_C_API_Issues.md (hunk 14 above)
Comment (zhaixuejun1993) [reply src/bindings/c/docs/How_to_Debug_C_API_Issues.md]: > ruby?

misunderstanding will delete. https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks#syntax-highlighting

File: src/bindings/c/docs/how_to_write_unite_test.md [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+| Unite Case Component | Location | Description |
+|:---     |:---   |:---
+|Core|[ov_core_test.cpp](../tests/ov_core_test.cpp)| including all core related interfaces tests
+|Model|[ov_model_test.cpp](../tests/ov_model_test.cpp)| including all model related interfaces tests
+|Compiled Model|[ov_compiled_model_test.cpp](../tests/ov_compiled_model_test.cpp)| including all compiled model related interfaces tests
+|Infer Request|[ov_infer_request_test.cpp](../tests/ov_infer_request_test.cpp)| including all infer request related interfaces tests
+|Tensor|[ov_tensor_test.cpp](../tests/ov_tensor_test.cpp)| including all tensor related interfaces tests
+|Partial Shape|[ov_partial_shape_test.cpp](../tests/ov_partial_shape_test.cpp)| including all partial shape related interfaces tests
+|Layout|[ov_layout_test.cpp](../tests/ov_layout_test.cpp)| including all layout related interfaces tests
+|Preprocess|[ov_preprocess_test.cpp](../tests/ov_preprocess_test.cpp)| including all preprocess related interfaces tests
```
Comment (ilyachur) [inline src/bindings/c/docs/how_to_write_unite_test.md:14]: I think we have a good file names which describe purpose for each file. I am not sure that we need to have this table. Just we can mention that all tests are located in the folder.

File: src/bindings/c/docs/how_to_write_unite_test.md [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+Here is an example for C interface unite test case:
+* C++ interface for read model,
+
+https://github.com/openvinotoolkit/openvino/blob/d96c25844d6cfd5ad131539c8a0928266127b05a/src/inference/include/openvino/runtime/core.hpp#L87-L100 
+
+* C wrap this interface like,
+
+https://github.com/openvinotoolkit/openvino/blob/d96c25844d6cfd5ad131539c8a0928266127b05a/src/bindings/c/src/ov_core.cpp#L71-L90
+
+* Create unite test case for this interface. At first, this interface is for core operation so the location should at [ov_core_test.cpp](../tests/ov_core_test.cpp). Also, the interface has default parameter so need to make unite test case for parameter missing. The final based function level test like:
```
Comment (ilyachur) [inline src/bindings/c/docs/how_to_write_unite_test.md:27]: Good section. But from my point of view first two points are about: how to wrap C++ interface, not about How to write unit tests.

File: src/bindings/c/docs/how_to_write_unite_test.md [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+
+https://github.com/openvinotoolkit/openvino/blob/d96c25844d6cfd5ad131539c8a0928266127b05a/src/bindings/c/tests/ov_core_test.cpp#L39-L63
+
+
+
```
Comment (ilyachur) [inline src/bindings/c/docs/how_to_write_unite_test.md:32]: Add the end of each file will be great to have references to some related pages (like in common readme)

File: src/bindings/c/docs/how_to_wrap_openvino_objects_with_c.md [hunk 23]
```diff
@@ -0,0 +1,88 @@
...
+
+Tips:
+1) For the objects which needs to be hided for users, C `struct` contains a shared pointer will be adopted.
+2) For the objects which needs to be created, operated and read by users, rewrite the C++ `class` will be better.
+3) For some simple objects, C `struct` contains a instance of C++ `class` will be enough.
+
+In OpenVINO C++, most objects implemented with C++ `class` as following:
+
+```
+class ClassName {
```
Comment (ilyachur) [inline src/bindings/c/docs/how_to_wrap_openvino_objects_with_c.md:18]: Looks like we shouldn't provide the base principles of C++ language here :) Developer already should know it

File: src/bindings/c/docs/how_to_wrap_openvino_objects_with_c.md [hunk 24]
```diff
@@ -0,0 +1,88 @@
...
+C provides the `struct` by `typedef struct ov_layout ov_layout_t;`
+
+ ## Wrap by Rewrite
+
+C construct a new `struct` represents the class, which rewrites related info to the `class` as following:
+
+```
+typedef struct {
+    ov::ClassName object;
+} ov_ClassName_t;
```
Comment (ilyachur) [inline src/bindings/c/docs/how_to_wrap_openvino_objects_with_c.md:75]: We don't use CamelCase in C API

File: src/bindings/c/docs/how_to_wrap_openvino_objects_with_c.md [hunk 25]
```diff
@@ -0,0 +1,88 @@
...
+https://github.com/openvinotoolkit/openvino/blob/d96c25844d6cfd5ad131539c8a0928266127b05a/src/bindings/c/src/common.h#L47-L53
+
+C provides the `struct` by `typedef struct ov_core ov_core_t;`
+
+ ## Wrap by C++ Object
+
+C construct a new `struct` represents the class, which contains an instance to C++ `class` as following:
+
+```
+struct ov_ClassName {
```
Comment (ilyachur) [inline src/bindings/c/docs/how_to_wrap_openvino_objects_with_c.md:53]: We don't use CamelCase in C API

File: src/bindings/c/docs/how_to_wrap_openvino_objects_with_c.md [hunk 26]
```diff
@@ -0,0 +1,88 @@
...
+https://github.com/openvinotoolkit/openvino/blob/d96c25844d6cfd5ad131539c8a0928266127b05a/src/core/include/openvino/core/shape.hpp#L21-L40
+
+Represent by C `struct` [here](../src/common.h)
+
+https://github.com/openvinotoolkit/openvino/blob/d96c25844d6cfd5ad131539c8a0928266127b05a/src/bindings/c/include/openvino/c/ov_shape.h#L15-L22
+
+> **NOTE**: this implementation needs developer create the C++ `class` based on the C `struct` info in using this rewrite type.
+
+ ## See also
+ * [Mapping Relationship of Objects](./docs/mapping_relationship_of_objects.md)
```
Comment (ilyachur) [inline src/bindings/c/docs/how_to_wrap_openvino_objects_with_c.md:87]: Add reference to developer documentation.

File: src/bindings/c/docs/how_to_wrap_openvino_interfaces_with_c.md [hunk 27]
```diff
@@ -0,0 +1,81 @@
...
+
+As we can see no C++ interface called. But when the object needs to be used for C++ call, which will provide info to create the related C++ object before call, such as:
+
+https://github.com/openvinotoolkit/openvino/blob/d96c25844d6cfd5ad131539c8a0928266127b05a/src/bindings/c/src/ov_tensor.cpp#L41-L55
+
+The tensor create needs to specify the shape info, so C shape need to be converted to C++ class before C++ interface using. Vice versa, the object needs to convert to C rewriting object for C users convenience.
+
+ ## See also
+ * [Mapping Relationship of Objects](./docs/mapping_relationship_of_objects.md)
+ * [Mapping Relationship of Interfaces](./docs/mapping_relationship_of_interfaces.md).
```
Comment (ilyachur) [inline src/bindings/c/docs/how_to_wrap_openvino_interfaces_with_c.md:79]: Add references to up-level developer documents

File: src/bindings/c/docs/how_to_debug_c_api_issues.md [hunk 28]
```diff
@@ -0,0 +1,13 @@
...
+
+https://github.com/openvinotoolkit/openvino/blob/d96c25844d6cfd5ad131539c8a0928266127b05a/src/bindings/c/include/openvino/c/ov_common.h#L68-L96
+
+As known, C API is a wrapping for C++ API, the main issues we had met can be classified into two kinds. Issue from input parameters checking and from C++ call exception.
+* for parameter checking issue: return value is -14, please check the input parameters
+* for C++ call exception issue: C interface just returns the status value, no more detail info message. If you want details please print it in exception macro.
+
+> **NOTE**: The exception from C interface is not the same as C++ exception, do not use the C status value in C++ debuging.
+
+
```
Comment (ilyachur) [inline src/bindings/c/docs/how_to_debug_c_api_issues.md:13]: Add references to other documents

File: src/bindings/c/docs/how_to_debug_c_api_issues.md [hunk 29]
```diff
@@ -0,0 +1,13 @@
...
+C API provides exception handling, here is all possible return values of the functions:
+
+https://github.com/openvinotoolkit/openvino/blob/d96c25844d6cfd5ad131539c8a0928266127b05a/src/bindings/c/include/openvino/c/ov_common.h#L68-L96
+
+As known, C API is a wrapping for C++ API, the main issues we had met can be classified into two kinds. Issue from input parameters checking and from C++ call exception.
+* for parameter checking issue: return value is -14, please check the input parameters
+* for C++ call exception issue: C interface just returns the status value, no more detail info message. If you want details please print it in exception macro.
+
+> **NOTE**: The exception from C interface is not the same as C++ exception, do not use the C status value in C++ debuging.
```
Comment (ilyachur) [inline src/bindings/c/docs/how_to_debug_c_api_issues.md:11]: I think we also can add link to common debug capabilities

File: src/bindings/c/docs/mapping_relationship_of_objects.md [hunk 30]
```diff
@@ -0,0 +1,30 @@
+# Mapping Relationship of Objects
+
+Here is the details introduction about the matching relationship between C structures and C++ classes: 
+| C Structure | C++ Class | Description | File Location
```
[the remark this answers is not in the document]
Comment (riverlijunjie) [reply src/bindings/c/docs/mapping_relationship_of_objects.md:4]: This table provides more clear view between C and C++, but it is not a must information/description.

File: src/bindings/c/docs/samples_implemented_by_openvino_c_api.md [hunk 31]
```diff
@@ -0,0 +1,6 @@
+# Two Simple Demos implemented by OpenVINO C API
```
[the remark this answers is not in the document]
Comment (zhaixuejun1993) [reply src/bindings/c/docs/samples_implemented_by_openvino_c_api.md:1]: removed

File: src/bindings/c/docs/how_to_write_unite_test.md (hunk 20 above)
Comment (zhaixuejun1993) [reply src/bindings/c/docs/how_to_write_unite_test.md:14]: removed

File: src/bindings/c/docs/mapping_relationship_of_objects.md (hunk 30 above)
[the remark this answers is not in the document]
Comment (zhaixuejun1993) [reply src/bindings/c/docs/mapping_relationship_of_objects.md:4]: I think this Table provides a overview picture for developer, API reference is more detail for the further learning. So I suggest to keep it

File: src/bindings/c/docs/how_to_write_unite_test.md (hunk 21 above)
Comment (zhaixuejun1993) [reply src/bindings/c/docs/how_to_write_unite_test.md:27]: will use reference replace this two poins

File: src/bindings/c/docs/how_to_write_unite_test.md (hunk 22 above)
Comment (zhaixuejun1993) [reply src/bindings/c/docs/how_to_write_unite_test.md:32]: done

File: src/bindings/c/docs/how_to_wrap_openvino_objects_with_c.md (hunk 23 above)
Comment (zhaixuejun1993) [reply src/bindings/c/docs/how_to_wrap_openvino_objects_with_c.md:18]: removed

File: src/bindings/c/docs/how_to_wrap_openvino_objects_with_c.md (hunk 25 above)
Comment (zhaixuejun1993) [reply src/bindings/c/docs/how_to_wrap_openvino_objects_with_c.md:53]: fixed

File: src/bindings/c/docs/how_to_debug_c_api_issues.md (hunk 28 above)
Comment (zhaixuejun1993) [reply src/bindings/c/docs/how_to_debug_c_api_issues.md:13]: done

File: src/bindings/c/docs/how_to_debug_c_api_issues.md (hunk 29 above)
Comment (zhaixuejun1993) [reply src/bindings/c/docs/how_to_debug_c_api_issues.md:11]: done

File: src/bindings/c/docs/how_to_wrap_openvino_objects_with_c.md (hunk 24 above)
Comment (zhaixuejun1993) [reply src/bindings/c/docs/how_to_wrap_openvino_objects_with_c.md:75]: fixed

**Review by ilyachur — COMMENTED** [review]
@tsavina Could you take a look?

File: src/bindings/c/docs/mapping_relationship_of_interfaces.md [hunk 32]
```diff
@@ -0,0 +1,136 @@
...
+ - [The mapping relationship for tensor](#the-mapping-relationship-for-tensor)
+ - [See also](#see-also)
+
+## The mapping relationship for core
+
+Here is the mapping relationship for core related,
+ * All core related APIs implemented with C++ can be found [here](../../../inference/include/openvino/runtime/core.hpp)
+ * All core related APIs implemented with C can be found [here](../include/openvino/c/ov_core.h)
+
+| C++ API | C API | Description |
```
Comment (ilyachur) [inline src/bindings/c/docs/mapping_relationship_of_interfaces.md:24]: These tables are hard to maintain.
Shouldn't Reference API documentation answer on the same questions?

File: src/bindings/c/docs/how_to_wrap_openvino_interfaces_with_c.md [hunk 33]
```diff
@@ -0,0 +1,81 @@
+# How to wrap OpenVINO interfaces with C
+
+Library `openvino_c` implements the most C++ interfaces provided in OpenVINO with C, the details mapping relationship can be found in [Mapping Relationship of Interfaces](./docs/mapping_relationship_of_interfaces.md). Here is the details about how to wrap interfaces(C++) from OpenVINO to interfaces(C), main steps including:
+ 1) define C interface name and input parameters
+ 2) check the validation of input parameters
+ 3) convert C input parameters to C++ interface using
+ 4) call C++ interface and get return
+ 5) convert C++ return object to C object
```
Comment (tsavina) [inline src/bindings/c/docs/how_to_wrap_openvino_interfaces_with_c.md:8]:
```suggestion
 1) Define C interface name and input parameters.
 2) Check the validation of input parameters.
 3) Convert C input parameters to C++ interface using.
 4) Call C++ interface and get return.
 5) Convert C++ return object to C object.
```

File: src/bindings/c/docs/how_to_wrap_openvino_interfaces_with_c.md [hunk 34]
```diff
@@ -0,0 +1,81 @@
...
+
+        cpp_parameter_convert-->c_return_success
+        c_wrap_call_cpp-->c_return_exception
+    end
+```
+
+According to the ability, all C provided interfaces can be classified to three kinds of methods:
+- Wrap C++ interface to create object
+- Wrap C++ interface to operate object
+- C implement interfaces
```
Comment (tsavina) [inline src/bindings/c/docs/how_to_wrap_openvino_interfaces_with_c.md:42]: kinda lost here does it mean Implement C interfaces or C-implemented interfaces?

File: src/bindings/c/docs/how_to_wrap_openvino_interfaces_with_c.md [hunk 35]
```diff
@@ -0,0 +1,81 @@
...
+        c_wrap_call_cpp-->c_return_exception
+    end
+```
+
+According to the ability, all C provided interfaces can be classified to three kinds of methods:
+- Wrap C++ interface to create object
+- Wrap C++ interface to operate object
+- C implement interfaces 
+
+## Wrap C++ interface to create object
```
Comment (tsavina) [inline src/bindings/c/docs/how_to_wrap_openvino_interfaces_with_c.md:44]:
```suggestion
## Wrap C++ interface to create an object
```

File: src/bindings/c/docs/mapping_relationship_of_objects.md (hunk 30 above)
[the remark this answers is not in the document]
Comment (zhaixuejun1993) [reply src/bindings/c/docs/mapping_relationship_of_objects.md]: removed the mapping relationship

File: src/bindings/c/docs/how_to_write_unite_test.md [hunk 36]
```diff
@@ -0,0 +1,16 @@
+# How to Write Unite Test for C API
```
Comment (ilyachur) [inline src/bindings/c/docs/how_to_write_unite_test.md:1]: Please fix the name `unit`

File: src/bindings/c/docs/how_to_write_unite_test.md [hunk 37]
```diff
@@ -0,0 +1,16 @@
...
+To ensure the accuracy of C API, all interfaces need to implement function level unite test at least. According to the object define, all implemented unite test cases are located in this folder.
+
+If developer wrap new interfaces from OpenVINO C++, you also need to add the unite test case in the correct location.
+Here is an example wrap C++ interface to C [wrap core](./how_to_wrap_openvino_interfaces_with_c.md).
+
+Create unite test case for this interface. At first, this interface is for core operation so the location should at [ov_core_test.cpp](../tests/ov_core_test.cpp). Also, the interface has default parameter so need to make unite test case for parameter missing. The final based function level test like:
+
+https://github.com/openvinotoolkit/openvino/blob/d96c25844d6cfd5ad131539c8a0928266127b05a/src/bindings/c/tests/ov_core_test.cpp#L39-L63
+
+## See also
```
Comment (ilyachur) [inline src/bindings/c/docs/how_to_write_unite_test.md:12]: Please add the information about tests binary file name

File: src/bindings/c/README.md [hunk 38]
```diff
@@ -0,0 +1,51 @@
...
+> **NOTE**:  Using API 2.0 is strongly recommended. Legacy API (for C) [header file](./include/c_api/ie_c_api.h), [source file](./src/ie_c_api.cpp), [unite test](./tests/ie_c_api_test.cpp) are also included in the component, but the legacy API is no longer extended. 
+
+## Tutorials
+
+* [How to integrate OpenVINO C API with Your Application](https://docs.openvino.ai/latest/openvino_docs_OV_UG_Integrate_OV_with_your_application.html)
+* [How to wrap OpenVINO objects with C](./docs/how_to_wrap_openvino_objects_with_c.md)
+* [How to wrap OpenVINO interfaces with C](./docs/how_to_wrap_openvino_interfaces_with_c.md)
+* [Samples implemented by OpenVINO C API](../../../samples/c/hello_classification/README.md)
+* [How to debug C API issues](./docs/how_to_debug_c_api_issues.md)
+* [How to write unit test](./docs/how_to_write_unit_test.md)
```
Comment (ilyachur) [inline src/bindings/c/README.md:41]: Incorrect link

File: src/bindings/c/docs/how_to_debug_c_api_issues.md [hunk 39]
```diff
@@ -0,0 +1,16 @@
...
+
+C API provides exception handling, here are all possible return values of the functions:
+
+https://github.com/openvinotoolkit/openvino/blob/d96c25844d6cfd5ad131539c8a0928266127b05a/src/bindings/c/include/openvino/c/ov_common.h#L68-L96
+
+There are two main types of possible issues:
+* parameter checking issue: return value is -14, check the input parameters
+* C++ call exception issue: C interface just returns the status value, without a detailed message. If you want details, print it in exception macro.
+
+> **NOTE**: The exception from C interface is not the same as C++ exception, do not use the C status value in C++ debugging.
```
Comment (ilyachur) [inline src/bindings/c/docs/how_to_debug_c_api_issues.md:11]: We shouldn't have exceptions from C interface at all

File: src/bindings/c/README.md [hunk 40]
```diff
@@ -0,0 +1,51 @@
...
+ * [tests](./tests) contains all tests for OpenVINO C APIs. [Learn more](./docs/how_to_write_unite_test.md).
+
+> **NOTE**:  Using API 2.0 is strongly recommended. Legacy API (for C) [header file](./include/c_api/ie_c_api.h), [source file](./src/ie_c_api.cpp), [unite test](./tests/ie_c_api_test.cpp) are also included in the component, but the legacy API is no longer extended. 
+
+## Tutorials
+
+* [How to integrate OpenVINO C API with Your Application](https://docs.openvino.ai/latest/openvino_docs_OV_UG_Integrate_OV_with_your_application.html)
+* [How to wrap OpenVINO objects with C](./docs/how_to_wrap_openvino_objects_with_c.md)
+* [How to wrap OpenVINO interfaces with C](./docs/how_to_wrap_openvino_interfaces_with_c.md)
+* [Samples implemented by OpenVINO C API](../../../samples/c/hello_classification/README.md)
```
Comment (ilyachur) [inline src/bindings/c/README.md:39]:
```suggestion
* [Samples implemented by OpenVINO C API](../../../samples/c/)
```

**Review by ilyachur — APPROVED** [review]
@riverlijunjie Please take a look

**Comment by riverlijunjie** [comment]
> @riverlijunjie Please take a look

@ilyachur I have reviewed it, it looks good to me!

**Pull request merged**